src/LoanAgent.sol

```diff
@@ -49,73 +38,166 @@ contract LoanAgent {
         Terminated
     }
 
+    AddressInfo public addressInfo;
     NodeStatus public status;
+    LoanAgentInfo public loanAgentInfo;
+
+    /// checkpoint => withdrawnAmount
+    mapping(uint => uint) withdrawHistory;
+    /// checkpoint => available balance after withdrawal
+    mapping(uint => uint) availableBalanceHistory;
+    // checkpoint => miner share
+    mapping(uint => uint) ownerWithrawableAmount;
+    // total amount of miner share accummulated
+    uint totalOwnerCredit;
+    uint lastWithdrawlCheckpoint;
 
-    // Stakepool factory
     address public immutable factory;
+    address public immutable aFil;
 
-    // should be called using LoanAgentFactory
     constructor(
         AddressInfo memory _addressInfo,
         LoanAgentInfo memory _loanAgentInfo,
-        address _factory
+        address _factory,
+        address _aFil
     ) {
-        // Call ChangeWorkerAddress in MinerAPI: https://docs.zondax.ch/fevm/filecoin-solidity/api/actors/Miner#changeworkeraddress
+        addressInfo = _addressInfo;
+        loanAgentInfo = _loanAgentInfo;
         factory = _factory;
+        aFil = _aFil;
+        status = NodeStatus.InQueue;
     }
 
-    modifier onlyFactory {
+    modifier onlyFactory() {
+        require(msg.sender == factory, "can only be called by factory");
         _;
     }
 
-    modifier onlyOracle {
+    modifier onlyOracle() {
+        require(
+            msg.sender == addressInfo.oracle,
+            "can only be called by oracle"
+        );
         _;
     }
 
-    modifier onlyRealOwner {
+    modifier onlyOracleOrFactory() {
+        require(
+            msg.sender == addressInfo.oracle || msg.sender == factory,
```

Comment:
Their should be clear distinction between factory and oracle calls.

Reply:
it's used for the `updateStatus` function. Both Factory and Oracle contract need to call that function. And seems like I can't OR 2 modifiers.